docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,519 @@
+# ADR 008: RPC Client Event Subscription Mechanism
+
+## Changelog
+
+* 2020-07-23: Initial draft
+
+## Context
+
+The [Tendermint Light Client](../../light-client/) is one of many important
+applications that will make use of the [RPC client](../../rpc/) to query
+Tendermint full nodes for information relating to the blockchain.
+
+Tendermint servers (e.g. full nodes) provide an [event
+subscription][tm-event-subs] RPC endpoint to allow clients to receive
+notifications of specific events as they happen (currently via a WebSockets
+connection). We need to expose this subscription mechanism through the RPC client.
+
+In order to achieve this, we need:
+
+1. An ergonomic interface for the RPC client that allows developers to subscribe
+   to events produced by specific queries. Specifically, this interface must:
+   1. Offer **subscription** functionality, where:
+      1. A single **subscription** takes place to the **events** produced by a
+         **query** (the [PEG] for which, at the time of this writing, is located
+         [here][query-peg]).
+      2. Callers must be able to create multiple distinct subscriptions.
+      3. A subscription must offer an interface to allow for iteration over
+         incoming events only relevant to that subscription (i.e. it should
+         produce an **event stream**).
+      4. It must be possible to, from outside of the RPC client, receive events
+         from multiple subscriptions' event streams concurrently without
+         interfering with/blocking each other.
+      5. It must be possible to, from outside of the RPC client, handle
+         subscriptions' event streams without blocking other RPC operations.
```

Comment:
What about the ability to receive all subscriptions on a single stream? eg. if I want to subscribe to a bunch of distinct event types and receive them in order?

And would it make sense / simplify things for the base subscription system to be like that, ie. a single stream with all subscriptions, and have the multi streams (I guess the Router) be an optional layer on top?

Reply:
My understanding of the initial reasoning around having `Subscription` implement the [`Stream`](https://docs.rs/futures/0.3.5/futures/stream/trait.Stream.html) trait is that it opens up possibilities around the use of more general `futures` functionality like [`select_all`](https://docs.rs/futures/0.3.5/futures/stream/fn.select_all.html), which would allow your code to combine multiple subscription streams into a single one and iterate through them as one stream.

Reply:
Interesting, and that would preserve order from the underlying socket? Does that make more sense than an approach where all events on the websocket itself are a single Stream and clients have option of partitioning that up into possibly multiple streams per subscription?

Reply:
The docs for `select_all` claim that elements from the underlying streams are returned while preserving the order in which they are produced by the underlying streams. I can add some tests into #516 that verify this, because I haven't actually tested it yet.

In theory it should work like this. The WebSocket driver I've implemented in #516 pushes events out to subscriptions (into their internal `tokio::mpsc` buffers) immediately as they're received before trying to read another event from the WebSocket connection, so this approach should preserve ordering when using `select_all`.